### PR TITLE
cpython: several fixes

### LINKED ISSUE #10808: [package] cpython/3.10.0: Imported target includes non-existent path.
### Package and Environment Details

  * Package Name/Version: **cpython/3.10.0**
  * Operating System+version: **Windows 11**
  * Compiler+version: **MSVC 2019**
  * Docker image: **None**
  * Conan version: **conan 1.45.0**
  * Python version: **Python 3.9.2**

### Conan profile (output of `conan profile show default`)

```
[settings]
os=Windows
os_build=Windows
arch=x86_64
arch_build=x86_64
build_type=Release
[options]
[conf]
[build_requires]
[env]
```

### CMakeLists.txt and main.cpp

I use [CMake-Conan](https://github.com/conan-io/cmake-conan) to install package.

-   <details><summary>Click to expand the CMakeLists.txt</summary>

    ```cmake
    cmake_minimum_required(VERSION 3.15)
    get_filename_component(folder_name "${CMAKE_CURRENT_SOURCE_DIR}" NAME)
    project(${folder_name} LANGUAGES C CXX)
    message("========== ${folder_name} ==========")


    get_cmake_property(is_multi_config GENERATOR_IS_MULTI_CONFIG)
    if(is_multi_config)     # Multi-Config Generator
      set(output_dir "${CMAKE_BINARY_DIR}/$<CONFIG>")
    else(is_multi_config)   # Single-Config Generator
      set(output_dir "${CMAKE_BINARY_DIR}/${CMAKE_BUILD_TYPE}")
    endif(is_multi_config)
    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${output_dir}/bin")
    set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY "${output_dir}/lib")
    set(CMAKE_LIBRARY_OUTPUT_DIRECTORY "${output_dir}/lib")


    if(CMAKE_CXX_COMPILER_ID MATCHES MSVC)
      add_compile_options("/source-charset:utf-8")
    elseif(CMAKE_CXX_COMPILER_DIR MATCHES GNU)
      add_compile_options("-finput-charset=utf-8")
    endif()


    if(NOT EXISTS "${CMAKE_BINARY_DIR}/conan.cmake")
      message(STATUS "Downloading conan.cmake from https://github.com/conan-io/cmake-conan ...")
      file(DOWNLOAD 
        "https://raw.githubusercontent.com/conan-io/cmake-conan/develop/conan.cmake"
        "${CMAKE_BINARY_DIR}/conan.cmake")
    endif()
    include("${CMAKE_BINARY_DIR}/conan.cmake")


    set(conan_config_generators "CMakeDeps")
    set(conan_config_requires   "cpython/3.10.0")
    set(conan_config_options    "cpython:shared=True")


    if(is_multi_config)     # Multi-Config Generator
      foreach(type Debug;Release)
        conan_cmake_configure(
          GENERATORS  ${conan_config_generators}
          REQUIRES    ${conan_config_requires}
          OPTIONS     ${conan_config_options}
          IMPORTS     "bin, *.dll -> ../${type}/bin")
        conan_cmake_autodetect(settings BUILD_TYPE ${type})
        conan_cmake_install(
          PATH_OR_REFERENCE "."
          INSTALL_FOLDER    "./conan"
          BUILD             missing
          REMOTE            conancenter
          SETTINGS          ${settings})
      endforeach()
    else(is_multi_config)   # Single-Config Generator
      set(type ${CMAKE_BUILD_TYPE})
      conan_cmake_configure(
        GENERATORS  ${conan_config_generators}
        REQUIRES    ${conan_config_requires}
        OPTIONS     ${conan_config_options}
        IMPORTS     "bin, *.dll -> ../${type}/bin")
      conan_cmake_autodetect(settings)
      conan_cmake_install(
        PATH_OR_REFERENCE "."
        INSTALL_FOLDER    "./conan"
        BUILD             missing
        REMOTE            conancenter
        SETTINGS          ${settings})
    endif(is_multi_config)


    list(APPEND CMAKE_MODULE_PATH "${CMAKE_BINARY_DIR}/conan")
    list(APPEND CMAKE_PREFIX_PATH "${CMAKE_BINARY_DIR}/conan")

    find_package(android-ndk CONFIG REQUIRED)

    add_executable(main "./main.cpp")
    target_link_libraries(main PRIVATE cpython::cpython)
    ```

    </details>

-   <details><summary>Click to expand the main.cpp</summary>

    ```c++
    #include <iostream>
    using namespace std;
    int main()
    {
        cout << "Hello world" << endl;
        return 0;
    }
    ```

    </details>

### Steps to reproduce

1. Use VSCode to open an empty folder.
2. Download [CMake-Tools](https://github.com/microsoft/vscode-cmake-tools) plugin from Extension.
3. Create an empty `CMakeLists.txt` file, and then paste the above example on it.
4. Create an empty `main.cpp` file, and then paste the above example on it.
5. Press `Ctrl+Shift+P`, and then type `CMake: Configure`.
6. Press `Ctrl+Shift+P`, and then type `CMake: Build`.

### Logs

```
[cmake] CMake Error in CMakeLists.txt:
[cmake]   Imported target "cpython::cpython" includes non-existent path
[cmake] 
[cmake]     "C:/.conan/19e585/1/include"
[cmake] 
[cmake]   in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:
[cmake] 
[cmake]   * The path was deleted, renamed, or moved to another location.
[cmake] 
[cmake]   * An install or uninstall procedure did not complete successfully.
[cmake] 
[cmake]   * The installation package was faulty and references files it does not
[cmake]   provide.
```

### PR DESCRIPTION
**Pull request by maksim-petukhov**
Specify library name and version:  **cpython/\***

- Add Python 3.10.3
- Add option for restoring Windows 7 support
- Add ability to skip upgrading VS projects when `devenv` is not available (e.g. when using BuildTools)
- Disable linking system ncurses on Linux when `with_curses=False`
- Disable linking system tcl, tk on Linux when `with_tkinter=False`
- Fixes #10808

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/cpython/all/patches/3.10.3-0005-do-not-search-for-system-tkinter.patch [hunk 1]
```diff
@@ -0,0 +1,11 @@
+--- setup.py
++++ setup.py
+@@ -2106,6 +2106,8 @@
+         if self.detect_tkinter_fromenv():
+             return True
+ 
++        return False
++
+         # Rather than complicate the code below, detecting and building
```
Comment (madebr) [inline recipes/cpython/all/patches/3.10.3-0005-do-not-search-for-system-tkinter.patch:9]: Please modify this patch to keep the number of lines the same.
So remove an empty line and replace it with the `return False`.
Keeping the number of lines the same makes it easer to stack and re-order patches: the line offset remains the same.

File: recipes/cpython/all/conandata.yml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 sources:
-  "3.10.0":
-    url: "https://www.python.org/ftp/python/3.10.0/Python-3.10.0.tgz"
-    sha256: "c4e0cbad57c90690cb813fb4663ef670b4d0f587d8171e2c42bd4c9245bd2758"
+  "3.10.3":
+    url: "https://www.python.org/ftp/python/3.10.3/Python-3.10.3.tgz"
+    sha256: "5a3b029bad70ba2a019ebff08a65060a8b9b542ffc1a83c697f1449ecca9813b"
```
Comment (madebr) [inline recipes/cpython/all/conandata.yml:4]: Please don't remove older versions from the most recent minor version.
We will remove support for older minor releases when cpython is on 3.11 (like the cmake recipe does).

File: recipes/cpython/all/conanfile.py [hunk 3]
```diff
@@ -178,7 +183,7 @@ def requirements(self):
             self.requires("openssl/1.1.1l")
             self.requires("expat/2.4.1")
             if self._with_libffi:
-                self.requires("libffi/3.2.1")
+                self.requires("libffi/3.4.2")
```
[the remark this answers is not in the document]
Comment (maksim-0) [reply recipes/cpython/all/conanfile.py:186]: Why?

[the remark this answers is not in the document]
Comment (madebr) [reply recipes/cpython/all/conanfile.py:186]: If it's working, fine.
But in my experience with cpython + libffi, there is some coupling with libffi versioning and cpython versions.

**Comment by maksim-0** [comment]
Well, I see that cpython/3.9.7 Linux clang 11 build fails. Some strange error related to `lsb_release` execution. No idea what does it mean. Alright, let's reproduce it locally. I pulled **conanio/clang11-ubuntu16.04:latest** docker image, then mounted **cci** repo and run ` CONAN_SYSREQUIRES_MODE=enabled conan create . cpython/3.9.7@ -o cpython:shared=False -b "*"` and build succeeded (package_id is the same `6fed5d956bd9552d48ad60ea65e6cb9f660d4743`). I don't know how to fix it. Need some help

@madebr

**Comment by madebr** [comment]
Tbh, I also have no clue what is going wrong.
I think there is some problem with the mixing of the virtual environment on c3i and the newly built cpython.

Perhaps @jgsogo is able to create a reproducer since he has most(?) knowledge about the internals of c3i.

**Comment by SSE4** [comment]
@uilianries do you rememeber what was the problem with `lsb-release`?

**Comment by maksim-0** [comment]
btw, the same problem in #10706

**Comment by jgsogo** [comment]
clang 11 uses old docker images: `conanio/clang11:1.47.0` and, indeed, inside that image `lsb_release` is not available...

**Comment by uilianries** [comment]
Clang 10 and Clang 11 are fixed now: https://github.com/conan-io/conan-docker-tools/pull/435

I just relaunched this PR now.

**Comment by uilianries** [comment]
Well, cpython/3.8.12 works in my machine ™ 

I'm running the latest 'conanio/clang11:1.47.0' image available in hub.docker

Here is my log: [cpython_3.8.12.log](https://github.com/conan-io/conan-center-index/files/8780382/cpython_3.8.12.log)

To execute my steps:

```
docker pull conanio/clang11:1.47.0
cd conan-center-index/
git remote add maksim-0 https://github.com/maksim-0/conan-center-index.git
git fetch maksim-0 cpython_fix
git checkout cpython_fix
cd recipes/cpython/ 
docker run --rm -ti -v ${PWD}:/home/conan/project conanio/clang11:1.47.0
export CONAN_SYSREQUIRES_MODE=enabled
conan create . cpython/3.8.12@
```

The result is in the file above. Now, why is it failing ... Does Ci didn't update the consumed docker image yet?

**Comment by uilianries** [comment]
Good news! Linux images are now running, we needed some interval to sync all images I guess. I just re-executed again and cpython/3.8.12 passed on Linux. However, now Mac is the new failure.

**Comment by SSE4** [comment]
> Good news! Linux images are now running, we needed some interval to sync all images I guess. I just re-executed again and cpython/3.8.12 passed on Linux. However, now Mac is the new failure.

@uilianries can you check local builds on mac? especially, inspecting the `config.log` for more hints. maybe we should specify `--with-universal-archs=x86_64` or some flag to disable universal builds? we definetely don't want any universal builds for now, just plain x86_64 (single arch).

**Comment by uilianries** [comment]
Okay, the version **3.9.13** is working, so I'll open a separated PR, replacing it.

File: recipes/cpython/all/conanfile.py [hunk 4]
```diff
@@ -416,16 +442,22 @@ def build(self):
...
-
         self._patch_sources()
         if self.settings.compiler == "Visual Studio":
             self._msvc_build()
         else:
-            autotools = self._configure_autotools()
-            autotools.make()
+            env = {}
+            if self.settings.compiler == "apple-clang":  # how to compute the right value?
+                env["MACOSX_DEPLOYMENT_TARGET"] = "10.0"  # FIXME : clang: error: invalid version number in 'MACOSX_DEPLOYMENT_TARGET=11.6'
```
Comment (SSE4) [inline recipes/cpython/all/conanfile.py:451]:
```suggestion
                env["MACOSX_DEPLOYMENT_TARGET"] = "10.3"  # FIXME : clang: error: invalid version number in 'MACOSX_DEPLOYMENT_TARGET=11.6'
```

**Comment by maksim-0** [comment]
Hook fails with
`[HOOK - conan-center.py] post_package_info(): ERROR: [INCLUDE PATH DOES NOT EXIST (KB-H071)] Component cpython::cpython include dir 'include' is listed in the recipe, but not found in package folder. The include dir should probably be fixed or removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H071) `
I tried locally to empty includedirs
`self.cpp_info.components[self.name].includedirs = []`
but got
`ConanException: cpython/3.10.0 package_info(): Component name cannot be the same as the package name: 'cpython'`.
Using global `cpp_info.includedirs` is not allowed because we have components. How can this be fixed? Could you please help me, @SSE4?

**Comment by SSE4** [comment]
I am not that good with components. maybe @uilianries @danimtb @jgsogo can suggest.

**Comment by jgsogo** [comment]
TBH, I don't understand the error, component `cpython::cpython` doesn't exist. Maybe we are looking at the wrong place and it is a bug in the hook logic?

**Comment by ericLemanissier** [comment]
it's a bug in the hook yes. The current workaround is to create an empty `include` folder

**Comment by maksim-0** [comment]
@madebr
I have a couple of questions if you don't mind.

1. Hooks complain about long paths on Windows. Should I add `short_paths=True` to the recipe?

2. https://github.com/python/cpython/blob/1bc13df6281e72c6a5bbc5b2b6b17b95b829d14b/setup.py#L1078
Seems like CPython needs `readline` library but the recipe doesn't define it as a requirement and we use system one. Is it intentional?

**Pull request closed without merging**